### PR TITLE
[core] Support compaction metrics

### LINKED ISSUE #1999: [Feature] Support compaction metrics
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Subtask of [PIP-3](https://cwiki.apache.org/confluence/display/PAIMON/PIP-3%3A+Introduce+metrics+for+Paimon)

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by schnappi17**
### Purpose

Linked issue: close #1999

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/operation/metrics/CompactionMetrics.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+/** Metrics to measure a compaction. */
+public class CompactionMetrics {
+    private static final Logger LOG = LoggerFactory.getLogger(CompactionMetrics.class);
+    private static final int HISTOGRAM_WINDOW_SIZE = 10_000;
+    private static final String GROUP_NAME = "compaction";
+
+    private final MetricGroup metricGroup;
+
+    public CompactionMetrics(MetricRegistry registry, String tableName) {
+        this.metricGroup = registry.tableMetricGroup(GROUP_NAME, tableName);
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/operation/metrics/CompactionMetrics.java:38]: Why not use `bucketMetricGroup`? Compaction is a process happening in each bucket.

File: paimon-core/src/main/java/org/apache/paimon/operation/metrics/CompactionStats.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        return rewriteFileSize(compactBefore);
+    }
+
+    protected long getRewriteOutputFileSize() {
+        return rewriteFileSize(compactAfter);
+    }
+
+    protected long getRewriteChangelogFileSize() {
+        return rewriteFileSize(compactChangelog);
+    }
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/operation/metrics/CompactionStats.java:71]: Keep these values in class members. Don't calculate each time. Gauge values are fetched periodically by metrics system (for example Promethues) so we should avoid repeated calculation.

File: paimon-core/src/main/java/org/apache/paimon/compact/CompactTask.java [hunk 3]
```diff
@@ -19,28 +19,57 @@
...
+
+import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.Callable;
 
 /** Compact task. */
 public abstract class CompactTask implements Callable<CompactResult> {
 
     private static final Logger LOG = LoggerFactory.getLogger(CompactTask.class);
+    private CompactionMetrics metrics;
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/compact/CompactTask.java:38]: NIT: This member can be final. Also please add `@Nullable`.

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/MergeTreeCompactManager.java [hunk 4]
```diff
@@ -53,21 +56,25 @@ public class MergeTreeCompactManager extends CompactFutureManager {
     private final int numSortedRunStopTrigger;
     private final CompactRewriter rewriter;
 
+    private final CompactionMetrics metrics;
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/MergeTreeCompactManager.java:59]: Add `@Nullable`.

File: paimon-core/src/main/java/org/apache/paimon/table/sink/TableWriteImpl.java [hunk 5]
```diff
@@ -146,6 +156,18 @@ public void compact(BinaryRow partition, int bucket, boolean fullCompaction) thr
         write.compact(partition, bucket, fullCompaction);
     }
 
+    @Override
+    public TableWrite withMetricRegistry(MetricRegistry registry, BinaryRow partition, int bucket) {
+        write.withMetrics(
+                new CompactionMetrics(registry, tableName, getPartitionString(partition), bucket));
+        return this;
+    }
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/table/sink/TableWriteImpl.java:164]: One `TableWrite` (and also one `FileStoreWrite`) is responsible for writing all partitions and buckets of a table. You can't specify partitions and buckets here. Please check `AbstractFileStoreWrite` to understand how writers for different buckets are grouped together.

**Review by tsreaper — CHANGES_REQUESTED** [review]

File: paimon-core/src/main/java/org/apache/paimon/table/sink/TableWriteImpl.java (hunk 5 above)
Comment (schnappi17) [reply paimon-core/src/main/java/org/apache/paimon/table/sink/TableWriteImpl.java:164]: Yes, I've notice that when developing the code for bridging to flink metrics system, I'll update this asap. Thank you for reminding me~ @tsreaper

**Review by tsreaper — APPROVED** [review]

**Pull request merged**